File: library/python/monitoring/solo/objects/solomon/v2/graph.py
# -*- coding: utf-8 -*-
import jsonobject
import six

try:
    from urllib import urlencode
except ImportError:
    from urllib.parse import urlencode

from library.python.monitoring.solo.objects.solomon.v2.base import SolomonObject


class SolomonCluster:
    STABLE = "https://solomon.yandex-team.ru"  # TODO: move to global module
    PRESTABLE = "http://solomon-prestable.yandex.net"
    TESTING = "http://solomon-test.yandex.net"


class Selector(SolomonObject):
    name = jsonobject.StringProperty(default="")
    value = jsonobject.StringProperty(default="")


class Parameter(SolomonObject):
    name = jsonobject.StringProperty(default="")
    value = jsonobject.StringProperty(default="")


class Element(SolomonObject):
    title = jsonobject.StringProperty(default="")
    type = jsonobject.StringProperty(default="SELECTORS")
    selectors = jsonobject.SetProperty(Selector)
    expression = jsonobject.StringProperty(default="")
    link = jsonobject.StringProperty(default="")
    area = jsonobject.BooleanProperty(default=True)
    stack = jsonobject.StringProperty(default="")
    down = jsonobject.BooleanProperty(exclude_if_none=True)
    color = jsonobject.StringProperty(default="")
    yaxis = jsonobject.StringProperty(default="LEFT")
    transform = jsonobject.StringProperty(default="NONE")


class Graph(SolomonObject):
    __OBJECT_TYPE__ = "Graph"

    id = jsonobject.StringProperty(default="")
    project_id = jsonobject.StringProperty(name="projectId")
    name = jsonobject.StringProperty(default="")
    description = jsonobject.StringProperty(default="")

    parameters = jsonobject.SetProperty(Parameter)
    elements = jsonobject.SetProperty(Element)

    # Display tab
    graph_mode = jsonobject.StringProperty(name="graphMode",
                                           choices=["GRAPH", "PIE", "BARS", "DISTRIBUTION", "HEATMAP", "NONE"],
                                           default="GRAPH")
    secondary_graph_mode = jsonobject.StringProperty(name="secondaryGraphMode",
                                                     choices=["PIE", "BARS", "DISTRIBUTION", "HEATMAP", "NONE"],
                                                     default="PIE")
    # graphMode: GRAPH
    stack = jsonobject.BooleanProperty(default=None)
    normalize = jsonobject.BooleanProperty()
    # graphMode: PIE
    color_scheme = jsonobject.StringProperty(name="colorScheme", choices=["AUTO", "DEFAULT", "GRADIENT"],
                                             default="AUTO")
    green = jsonobject.StringProperty(default="")
    yellow = jsonobject.StringProperty(default="")
    red = jsonobject.StringProperty(default="")
    violet = jsonobject.StringProperty(default="")
    # graphMode: BARS, Y-axis (UI).
    scale = jsonobject.StringProperty(choices=["NATURAL", "LOG"])
    # Y-axis (UI). example: "numberFormat": "3|G"
    number_format = jsonobject.StringProperty(name="numberFormat", default="")
    # graphMode: HEATMAP. Heatmap thresholds
    green_value = jsonobject.StringProperty(name="greenValue", default="")
    yellow_value = jsonobject.StringProperty(name="yellowValue", default="")
    red_value = jsonobject.StringProperty(name="redValue", default="")
    violet_value = jsonobject.StringProperty(name="violetValue", default="")
    # Others
    interpolate = jsonobject.StringProperty(choices=["LINEAR", "LEFT", "RIGHT", "NONE"], default="LINEAR")
    drop_nans = jsonobject.BooleanProperty(name="dropNans", default=None)
    aggr = jsonobject.StringProperty(choices=["AVG", "MIN", "MAX", "LAST", "SUM"], default="AVG")

    # Transformation tab
    transform = jsonobject.StringProperty(
        choices=["DIFFERENTIATE", "DIFFERENTIATE_WITH_NEGATIVE", "INTEGRATE", "MOVING_AVERAGE", "MOVING_PERCENTILE",
                 "DIFF", "NONE"], default="DIFFERENTIATE")
    over_lines_transform = jsonobject.StringProperty(name="overLinesTransform",
                                                     choices=["NONE", "PERCENTILE", "WEIGHTED_PERCENTILE", "SUMMARY"])
    # overLinesTransform: PERCENTILE
    percentiles = jsonobject.StringProperty(default="")
    ignore_inf = jsonobject.BooleanProperty(name="ignoreInf")
    # Filter
    filter = jsonobject.StringProperty(choices=["NONE", "TOP", "BOTTOM"], default="NONE")
    filter_by = jsonobject.StringProperty(name="filterBy", choices=["AVG", "MIN", "MAX", "LAST", "SUM"], default="AVG")
    filter_limit = jsonobject.StringProperty(name="filterLimit", default="")
    # Downsampling
    downsampling = jsonobject.StringProperty(choices=["AUTO", "BY_INTERVAL", "BY_POINTS", "OFF"], default="BY_INTERVAL")
    downsampling_aggr = jsonobject.StringProperty(name="downsamplingAggr", choices=["DEFAULT", "AVG", "MIN", "MAX", "LAST", "SUM"],
                                                  default="AVG")
    grid = jsonobject.StringProperty(default="")
    # Others
    limit = jsonobject.StringProperty(default="")

    min = jsonobject.StringProperty(default="")
    max = jsonobject.StringProperty(default="")

    moving_window = jsonobject.StringProperty(name="movingWindow", default="")
    moving_percentile = jsonobject.StringProperty(name="movingPercentile", default="")
    max_points = jsonobject.IntegerProperty(name="maxPoints")
    hide_no_data = jsonobject.BooleanProperty(name="hideNoData")
    threshold = jsonobject.FloatProperty(default=None)
    version = jsonobject.IntegerProperty(name="version", required=True, default=0)

    def get_link(self, use_legend=True, urlencoded=True, parametrise=None):
        """
        Returns link without server name
        :param use_legend: show legend on graph or not
        :param parametrise: dict[str, str]
        parametrise can be use in dashboards for setting parameters:
        for example, parametrise = {"environment": "{{environment}}"}
        will use "environment" label value from dashboard parameters
        :return:
        """
        details = {p.name: p.value for p in self.parameters}
        details.update({
            "project": self.project_id,
            "graph": self.id,
        })
        if use_legend is not None:
            details["legend"] = "1" if use_legend else "0"
        if parametrise:
            for parameter, value in six.iteritems(parametrise):
                try:
                    next(filter(lambda p: p.name == parameter, self.parameters))
                except Exception:
                    raise Exception("Parameter \"{}\" not found for graph \"{}\"!".format(parameter, self.id))
                details[parameter] = value

        sorted_details = sorted(six.iteritems(details), key=lambda val: val[0])
        if urlencoded:
            return "?{0}".format(urlencode(sorted_details))
        return "?{0}".format("&".join(key + "=" + value for key, value in sorted_details))

    def get_dashboard_link(self, use_legend=True, parametrise=None):
        return self.get_link(use_legend=use_legend, parametrise=parametrise, urlencoded=False)

    def get_full_link(self, time_window=""):
        """
        :time_window: str
        example values - "1h", "1d", "1w", "1w1d1h"
        """
        return "{}/{}&b={}".format(SolomonCluster.STABLE, self.get_link(), time_window)

File: cloud/storage/core/tools/common/go/solomon/sdk/graph.go
package sdk

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
)

type GraphID = string

type GraphParameters struct {
	Name  string `json:"name"`
	Value string `json:"value"`
}

type GraphSelectors struct {
	Name  string `json:"name"`
	Value string `json:"value"`
}

type GraphElements struct {
	Title     string `json:"title"`
	Type      string `json:"type"`
	Stack     string `json:"stack"`
	Color     string `json:"color"`
	Link      string `json:"link"`
	Transform string `json:"transform"`
	Yaxis     string `json:"yaxis"`

	Selectors []GraphSelectors `json:"selectors"`
}

type Graph struct {
	ID        string `json:"id"`
	ProjectID string `json:"projectId"`
	Version   uint   `json:"version"`

	Name        string `json:"name"`
	Description string `json:"description"`

	Parameters []GraphParameters `json:"parameters"`
	Elements   []GraphElements   `json:"elements"`

	Aggr                string `json:"aggr"`
	BucketLabel         string `json:"bucketLabel"`
	ColorScheme         string `json:"colorScheme"`
	Downsampling        string `json:"downsampling"`
	DownsamplingAggr    string `json:"downsamplingAggr"`
	DownsamplingFill    string `json:"downsamplingFill"`
	Filter              string `json:"filter"`
	FilterBy            string `json:"filterBy"`
	FilterLimit         string `json:"filterLimit"`
	GraphMode           string `json:"graphMode"`
	Green               string `json:"green"`
	GreenValue          string `json:"greenValue"`
	Grid                string `json:"grid"`
	HideNoData          bool   `json:"hideNoData"`
	IgnoreInf           bool   `json:"ignoreInf"`
	IgnoreMinStepMillis bool   `json:"ignoreMinStepMillis"`
	Interpolate         string `json:"interpolate"`
	Limit               string `json:"limit"`
	Max                 string `json:"max"`
	MaxPoints           int    `json:"maxPoints"`
	Min                 string `json:"min"`
	MovingPercentile    string `json:"movingPercentile"`
	MovingWindow        string `json:"movingWindow"`
	Normalize           bool   `json:"normalize"`
	NumberFormat        string `json:"numberFormat"`
	OverLinesTransform  string `json:"overLinesTransform"`
	Percentiles         string `json:"percentiles"`
	Red                 string `json:"red"`
	RedValue            string `json:"redValue"`
	Scale               string `json:"scale"`
	SecondaryGraphMode  string `json:"secondaryGraphMode"`
	Transform           string `json:"transform"`
	Violet              string `json:"violet"`
	VioletValue         string `json:"violetValue"`
	Yellow              string `json:"yellow"`
	YellowValue         string `json:"yellowValue"`
}

type listGraphsItem struct {
	ID string `json:"id"`
}

type listGraphsResponse struct {
	Items []listGraphsItem `json:"result"`
	Page  listPage         `json:"page"`
}

func (s *solomonClient) listGraphs(
	ctx context.Context,
	pageNum uint,
) ([]GraphID, listPage, error) {

	url := s.url + "/" + s.projectID + "/graphs?pageSize=1000"
	url += "&page=" + fmt.Sprint(pageNum)

	body, err := s.executeRequest(ctx, "listGraphs", "GET", url, nil)
	if err != nil {
		return nil, listPage{}, err
	}

	r := &listGraphsResponse{}
	if err := json.Unmarshal(body, r); err != nil {
		return nil, listPage{}, fmt.Errorf("listGraphs. Unmarshal error: %w", err)
	}

	var result []string

	for _, item := range r.Items {
		result = append(result, item.ID)
	}

	return result, r.Page, nil
}

func (s *solomonClient) ListGraphs(
	ctx context.Context,
) ([]GraphID, error) {

	var result []string

	pageNum := uint(0)
	for {
		clusters, page, err := s.listGraphs(ctx, pageNum)
		if err != nil {
			return result, err
		}

		result = append(result, clusters...)
		pageNum = page.Current + 1

		if pageNum >= page.PagesCount {
			break
		}
	}

	return result, nil
}

func (s *solomonClient) GetGraph(
	ctx context.Context,
	graphID GraphID,
) (Graph, error) {

	url := s.url + "/" + s.projectID + "/graphs/" + graphID

	body, err := s.executeRequest(ctx, "GetGraph", "GET", url, nil)
	if err != nil {
		return Graph{}, err
	}

	r := &Graph{}
	if err := json.Unmarshal(body, r); err != nil {
		return Graph{}, fmt.Errorf("GetGraph. Unmarshal error: %w", err)
	}

	return *r, nil
}

func (s *solomonClient) AddGraph(
	ctx context.Context,
	graph Graph,
) (Graph, error) {

	url := s.url + "/" + s.projectID + "/graphs"

	payload, err := json.Marshal(graph)
	if err != nil {
		return Graph{}, err
	}

	body, err := s.executeRequest(
		ctx,
		"AddGraph",
		"POST",
		url,
		bytes.NewBuffer(payload),
	)
	if err != nil {
		return Graph{}, err
	}

	r := &Graph{}
	if err := json.Unmarshal(body, r); err != nil {
		return Graph{}, err
	}

	return *r, nil
}

func (s *solomonClient) UpdateGraph(
	ctx context.Context,
	graph Graph,
) (Graph, error) {

	url := s.url + "/" + s.projectID + "/graphs/" + graph.ID

	payload, err := json.Marshal(graph)
	if err != nil {
		return Graph{}, err
	}

	body, err := s.executeRequest(
		ctx,
		"UpdateGraph",
		"PUT",
		url,
		bytes.NewBuffer(payload),
	)
	if err != nil {
		return Graph{}, err
	}

	r := &Graph{}
	if err := json.Unmarshal(body, r); err != nil {
		return Graph{}, err
	}

	return *r, nil
}

func (s *solomonClient) DeleteGraph(
	ctx context.Context,
	graphID GraphID,
) error {

	url := s.url + "/" + s.projectID + "/graphs/" + graphID

	_, err := s.executeRequest(ctx, "DeleteGraph", "DELETE", url, nil)
	if err != nil {
		return err
	}

	return nil
}
